File: Python/2644_촌수계산.py
def getDgree(n, x, y, matrix, visit, d, r):
	visit[x] = True
	
	vector = []
	for i in range(n):
		if matrix[x][i] == True and visit[i] == False: #만약 부모 관계이고 방문한 적 없다면
			if i == y: # 만약 원하는 y값(목적지)에 도달했으면 깊이를 return하고 끝
				r = d+1
				return r
			else: # 벡터에 추가
				vector.append(i)
				
	while vector: #bfs로 구현해도 되지만 현재의 깊이 계속적으로 반영 위해 재귀로 구현
		tmp = vector.pop()
		r = getDgree(n, tmp, y, matrix, visit, d+1, r)
	
	return r #목적지에 도달하지 않으면 -1이 계속 return될 것(만약 도달한 것이 있다면, 도달한 깊이 값 계속 return될 것)
	
if __name__ == "__main__":
	n = int(input())
	
	p = input().split()
	p = [int(i)-1 for i in p]
	
	m = int(input())
	
	matrix = [[False]*n for i in range(n)]
	for i in range(m):
		xy = input().split()
		xy = [int(i)-1 for i in xy]
		matrix[xy[0]][xy[1]] = True
		matrix[xy[1]][xy[0]] = True
	
	visit = [False]*n
	print(getDgree(n, p[0], p[1], matrix, visit, 0, -1))

File: C++/2644_촌수계산.cpp
#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

int getDegree(int n, int p1, int p2, bool **matrix, bool *visit, int d, int r);

int main() {
	int n;
	cin >> n;
	
	int p1, p2;
	cin >> p1 >> p2;
	
	bool **matrix = new bool*[n];
	for(int i = 0; i < n; i++){
		matrix[i] = new bool[n];
		memset(matrix[i],false,sizeof(bool)*n);
	}
	
	int m;
	cin >> m;
	for(int i = 0; i < m; i++){
		int x, y;
		cin >> x >> y;
		
		matrix[x-1][y-1] = true;
		matrix[y-1][x-1] = true;
	}
	
	bool *visit = new bool[n];
	cout << getDegree(n, p1-1, p2-1, matrix, visit, 0, -1);
	
	return 0;
}

int getDegree(int n, int p1, int p2, bool **matrix, bool *visit, int d, int r){
	visit[p1] = true;
	
	vector<int> v;
	for(int i = 0; i < n; i++){
		if(matrix[p1][i] == true && visit[i] == false){ //현재 사람과 부모 자식 관계이고 방문한 적 없는 노드 중
			if(p2 == i){ // 목적지에 도달할 수 있는 경우 존재한다면
				r = d+1; //트리 깊이+1을 return
				return r;
			}
			else{
				v.push_back(i);
			}
		}
	}
	while(!v.empty()){ //아니라면 다음 경우 살펴보기
		int tmp;
		tmp = v.back();
		v.pop_back();
		r = getDegree(n, tmp, p2, matrix, visit, d+1, r);
	}
	return r;
}